chore: migrate `react-router-dom` from v5 to v6

keita-determined opened the pull request:
## Description

[DET-7851]
<!---
Lead with the intended commit body in this description field. For breaking changes,
please include "BREAKING CHANGE:" at the beginning of your commit body.
At a minimum, this section should include a bracketed reference to the Jira ticket,
e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.
-->


## Test Plan
- Check all pages if router works as expected
- Check code if there's something I miss
<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions about this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

I followed [this instruction](https://github.com/remix-run/react-router/discussions/8753) and [this](https://reactrouter.com/docs/en/v6/upgrading/v5)

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-7851]: https://determinedai.atlassian.net/browse/DET-7851?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

keita-determined commented on webui/react/src/components/NotesCard.tsx:
```diff
@@ -135,7 +135,7 @@ const NotesCard: React.FC<Props> = (
         />
       </Spinner>
-      <Prompt
+      {/* <Prompt
```
https://github.com/remix-run/react-router/issues/8139
`Prompt` is not available in v6

keita-determined commented on webui/react/src/pages/InteractiveTask.tsx:
```diff
@@ -85,25 +84,25 @@ export const InteractiveTask: React.FC = () => {
-            name={taskName}
-            resourcePool={taskResourcePool}
-            type={taskType}
+            id={taskId!}
```
im thinking to `assert` these cuz using `??` is not appropriate. Any other better options?

hkang1 replied:
**question:** correct me if I'm wrong, but there's no guarantee that `taskId` is always defined right? you can still go to routes like `/wait/notebook/` (instead of `/wait/notebook/<some-id>`. Variables coming through URL params should probably be treated as conditional definition. So `??` seems appropriate, or not showing the respective component if it requires the prop to be defined...

keita-determined replied:
It used be like [this](https://github.com/determined-ai/determined/blob/2da2ed2a1f11a87bd5999200561671149bd10cc9/webui/react/src/pages/InteractiveTask.tsx#L12). I just followed the previous style which doesn't have `?`.

keita-determined commented on webui/react/src/index.tsx:
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
-import { Router } from 'react-router-dom';
+import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
```
This code should be safe because I set `history` lib to be the same version as `react-router-dom`
https://reactrouter.com/docs/en/v6/routers/history-router
https://github.com/remix-run/react-router/blob/f16c5490dfa75f15dcfb86d2a981a7c58a9d1a33/package.json#L44

hkang1 replied:
**suggestion:** can you add this comment above this line just so people reading this part of the code can lookup why it's called `unstable_HistoryRouter`
```
/**
 * It's considered unstable until `react-router-dom` can detect
 * history version mismatches when supplying your own history.
 * https://reactrouter.com/docs/en/v6/routers/history-router
 */
```

keita-determined commented on webui/react/package.json:
```diff
@@ -28,7 +28,7 @@
     "fuse.js": "^6.4.6",
     "hermes-parallel-coordinates": "^0.5.18",
-    "history": "^4.10.1",
+    "history": "^5.2.0",
```
https://github.com/remix-run/history/releases/tag/v5.0.0
there is breaking changes but In my investigation, theres no affects.

trentwatt commented:
From the instructions you linked:
>Instead of upgrading and updating all of your code at once (which is incredibly difficult and prone to bugs), this strategy enables you to upgrade one component, one hook, and one route at a time by running both v5 and v6 in parallel. Any code you haven't touched is still running the very same code it was before. Once all components are exclusively using the v6 APIs, your app no longer needs the compatibility package and is running on v6.

Where are we in this process? Does this PR just lay the groundwork for us to be able to upgrade components one by one? Or does it update some (or all??) of the components?

trentwatt commented:
@hkang1 

This one (or at least the migration PRs going forward) probably needs a pretty comprehensive test plan. Can we put together some summary spec of the existing (intended) behavior for navigation?

e.g. from project details, click on experiment. then click a tab. existing behavior in this case is that you go back to project details. is this desired?

also: `useSettings` has `shouldPush` param, which to my understanding means that the url-encoded setting gets it own entry in the navigation history-- is that right? different test cases involving url-encoding of settings without `shouldPush` probably also need to be incorporated as well.

there have also occasionally been issues around navigating e.g, `/det/resourcepool/compute-pool` and /det/resourcepool/aux-pools. 

also redirecting from tensorboard, opening jupyterlab in new window. navigating back from embedded view.

anything else come to mind?

hkang1 commented on webui/react/src/hooks/useSettings.ts:
```diff
@@ -1,6 +1,7 @@
 import queryString from 'query-string';
 import { useCallback, useEffect, useMemo, useState } from 'react';
-import { useHistory, useLocation } from 'react-router-dom';
+import { useLocation } from 'react-router-dom';
+import { useNavigate } from 'react-router-dom';
```
**nit:** combine into one line

hkang1 commented on webui/react/src/pages/ExperimentDetails/ExperimentSingleTrialTabs.tsx:
```diff
@@ -1,6 +1,7 @@
 import { Tabs } from 'antd';
 import React, { useCallback, useEffect, useState } from 'react';
-import { useHistory, useParams } from 'react-router';
+import { useParams } from 'react-router-dom';
+import { useNavigate } from 'react-router-dom';
```
**nit:** one line, maybe IDE needs adjustment to do this?

keita-determined replied:
I forgot to combine them when I replace `react-router-dom-v5-compat` with `react-router-dom`. Nice catch!

hkang1 commented on webui/react/src/pages/ExperimentDetails/ExperimentVisualization.tsx:
```diff
@@ -1,6 +1,7 @@
 import { Alert, Tabs } from 'antd';
 import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
-import { useHistory, useLocation } from 'react-router-dom';
+import { useLocation } from 'react-router-dom';
+import { useNavigate } from 'react-router-dom';
```
ditto

keita-determined commented:
> From the instructions you linked:
> 
> > Instead of upgrading and updating all of your code at once (which is incredibly difficult and prone to bugs), this strategy enables you to upgrade one component, one hook, and one route at a time by running both v5 and v6 in parallel. Any code you haven't touched is still running the very same code it was before. Once all components are exclusively using the v6 APIs, your app no longer needs the compatibility package and is running on v6.
> 
> Where are we in this process? Does this PR just lay the groundwork for us to be able to upgrade components one by one? Or does it update some (or all??) of the components?

Done with the entire process, so now on v6 completely. I tried to to update it little by little but our website relies on `history` library a lot instead of using `NavLink`. Although we don't need to have `history` library since `react-router-dom v6` have it as a dependency, we rely on `history` a lot, so I needed to handle differently which unables to co-exist both v5 and v6.
For example, [This](https://github.com/determined-ai/determined/pull/4619/files#diff-d211b226e853bf7f7d109f740456cb93344b177185a22a35c095610e4b066f72R2) syntax is changed. [New HistoryRouter](https://github.com/determined-ai/determined/pull/4619/files#diff-7ce3dfc51c01b4d5a4972c9d0770cfbe49a60d20e1c440cf755b13001843a53cR14) requires new version of `history`.
That's why I updated all from v5 and v6 at once. However, as @trentwatt said, we need to have comprehensive test cases to upgrade safely. I think I (or someone else) should write tests first, and then test this PR would be good.

hkang1 commented on webui/react/src/pages/InteractiveTask.tsx:
```diff
@@ -9,7 +9,7 @@ import { CommandState, CommandType } from 'types';
 import TaskLogs from './TaskLogs';
 
-interface Params {
+type Params = {
```
**question:** related to the question written below, should `taskId`, `taskName` and `taskResourcePool` be written with the `?` suffix (e.g. `taskId?: string;`)?

much like `ProjectDetails.tsx Params`

keita-determined replied:
in `react-router-dom v6`, `useParam` returns always `string | undefined` even when we don't put `?`. 
In v5, `useParam` returns only string if we dont put `?`, so it used be like [this](https://github.com/determined-ai/determined/blob/2da2ed2a1f11a87bd5999200561671149bd10cc9/webui/react/src/pages/InteractiveTask.tsx#L12). I just followed the previous style which doesnt have `?`. 
Also, I replaced `interface` with `type` due to type error in a new version.

hkang1 commented on webui/react/src/pages/TaskLogs.tsx:
```diff
@@ -16,7 +16,7 @@ import { CommandType } from 'types';
 import css from './TaskLogs.module.scss';
 
-interface Params {
+type Params = {
```
ditto: conditional params?

hkang1 commented on webui/react/src/pages/TrialDetails.tsx:
```diff
@@ -1,6 +1,7 @@
 import { Tabs } from 'antd';
 import React, { useCallback, useEffect, useRef, useState } from 'react';
-import { useHistory, useParams } from 'react-router';
+import { useParams } from 'react-router-dom';
+import { useNavigate } from 'react-router-dom';
```
ditto

hkang1 commented on webui/react/src/pages/Wait.tsx:
```diff
@@ -17,7 +17,7 @@ import { WaitStatus } from 'wait';
 import css from './Wait.module.scss';
 
-interface Params {
+type Params = {
```
ditto: conditional params? unless the router is preventing this page from rendering if the params are not provided?

keita-determined replied:
i use `assertIsDefined` to check if `undefined` or not. Even if the variable is undefined, the page will redirect to default page, and `assertIsDefined` doesnt actually throw error

hkang1 commented on webui/react/src/pages/WorkspaceList.tsx:
```diff
@@ -248,6 +248,7 @@ const WorkspaceList: React.FC = () => {
               limit: settings.tableLimit,
               offset: settings.tableOffset,
             }, total)}
+            rowKey="id"
```
👍

hkang1 commented on webui/react/src/routes/routes.ts:
```diff
@@ -113,21 +125,41 @@ const routes: RouteConfig[] = [
     id: 'clusters',
     needAuth: true,
-    path: '/clusters/:tab?',
+    path: '/clusters',
```
**question:** huh, is this something that changed with the version bump or was it always like this? I remember if we didn't put the more specific path before the general path in the array, it would resolve to the general before the more specific route, but playing with the preview, it doesn't seem to be the case?

keita-determined replied:
It comes from [this](https://github.com/remix-run/react-router/discussions/8753)
<img width="1392" alt="image" src="https://user-images.githubusercontent.com/109113616/180822015-c413b42b-8a6a-4a16-9e2c-64822fe9b9db.png">
`react-router-dom v6` doesnt support this syntax so I changed it

hkang1 replied:
it's a bit more verbose but it's cleaner IMO

hkang1 commented on webui/react/src/shared/routes/history.ts:
```diff
@@ -1,2 +1,2 @@
 import { createBrowserHistory } from 'history';
-export default createBrowserHistory({ basename: process.env.PUBLIC_URL });
+export default createBrowserHistory();
```
**question:** this is ok to do? same question with the changes in `stripUrl`

keita-determined replied:
The latest `history` library doesnt provide `basename` anymore, and we need to update `history` as well as `react-router-dom`

hamidzr replied:
some use cases that rely on PUBLIC_URL are:
- serve ui from any base path: eg `/det` `/` and `/cat`

keita-determined replied:
I moved base name to [here](https://github.com/determined-ai/determined/pull/4619/files#diff-fe9fa55e9afd85c5adb046f05a63357a7ed70bca08fae9622936c71c5a6f39b6R23). I tested most of pages manually, but we should write tests before this PR is merged.

keita-determined commented on webui/react/src/components/NotesCard.tsx:
```diff
@@ -135,18 +133,6 @@ const NotesCard: React.FC<Props> = (
           onClick={handleNotesClick}
         />
       </Spinner>
-      <Prompt
-        message={(newLocation) => {
-          const isSameExperiment = location.pathname.split('/')[0] === 'experiment' &&
-             newLocation.pathname.startsWith(location.pathname.split('/').slice(0, -1).join('/'));
-          return (
-            isSameExperiment ?
-              true :
-              'You have unsaved notes, are you sure you want to leave? Unsaved notes will be lost.'
-          );
-        }}
-        when={editedNotes !== notes}
-      />
```
I made a comment [here](https://github.com/determined-ai/determined/pull/4619#discussion_r927176853). 
It is already buggy before the update. I think we should find the alternative or wait for the new version which supports this.

hamidzr commented on webui/react/src/shared/utils/routes.ts:
```diff
@@ -48,9 +48,6 @@ export const parseUrl = (url: string): URL => {
 const stripUrl = (aUrl: string): string => {
   const url = parseUrl(aUrl);
   const rest = url.href.replace(url.origin, '');
-  if (rest.startsWith(process.env.PUBLIC_URL)) {
```
why were we using this 🤔

keita-determined replied:
not 100% sure, but its for netlify and production mode. In local, there's `/det` but on netlify there isn't.
@trentwatt might know about it more.

hamidzr replied:
most likely I wrote this.. do we need to change or remove it?

keita-determined replied:
the `history.ts` is changed, so the basename is not included anymore. That's why I removed `replace()` part.

hamidzr replied:
thanks. I think if [these](https://github.com/determined-ai/determined/pull/4619/files#r929350828) use cases still work we should be good here

trentwatt commented on webui/react/src/components/Router.tsx:
```diff
@@ -32,25 +33,23 @@ const Router: React.FC<Props> = (props: Props) => {
                 />
               )}
+              key={route.id}
+              {...route}
```
move to top?

keita-determined commented:
@hkang1 @hamidzr @trentwatt 

## New Changes Since the Last Review
- Found the alternative of `Prompt`
- Fix minor bugs
  - Order of props in `Router.tsx`
- Added `assertIsDefined` to check if params are defined (it is mostly for typescript type checking)
  - Even though the code has `throw Error`, it actually doesnt run since if the params are missing, page is redirected to the `paroject/1` or if the params do not exist, the page says `id xxx doesnt exist` or something similar 

## Concern
- In the current behavior with old `react-router-dom` and `history`, it automatically redirect to the default path if the path doesn't exist, such as `http://localhost:3000/` -> `http://localhost:3000/det`. However, due to the change of both `react-router-dom` and `history` (`history` doesnt take `basename` param anymore), we have to manually add `/det` if we run it locally.

trentwatt reviewed (commented):
I think redirecting `serverAddress` -> `serverAddress/det` is a behavior that needs to be preserved with any PR, unless we can confirm otherwise.

as of now, if i go to `localhost:3000` I get a blank page. I could see a situation where customers have an autocompleted  address that they go to, and if that ends up being a blank page, it would not be a great thing.

also i would prefer that we use typescript type assertions instead of writing our own `assertIsDefined` that throws and error (and crashes the page on my local)

Also couple bugs as discussed:
- clicking on project card results in blank page
- forking experiment results in blank page
- creating new project results in blank page
- creating new workspace results in blank page

probably some other things? can you verify all of the different pages and functionalities?

keita-determined commented:
> ## Concern
> * In the current behavior with old `react-router-dom` and `history`, it automatically redirect to the default path if the path doesn't exist, such as `http://localhost:3000/` -> `http://localhost:3000/det`. However, due to the change of both `react-router-dom` and `history` (`history` doesnt take `basename` param anymore), we have to manually add `/det` if we run it locally.

This concern is fixed!!

trentwatt reviewed (commented):
Couple more regressions: 

https://user-images.githubusercontent.com/28383080/184429853-fd693eae-5ecc-4f51-b568-a7048d536111.mov



- changing the password results in blank screen
- another time after changing password and going to `localhost:3000` I kept getting redirected to blank screen
- another time the page crashed immediately after login.

At this point I would prefer that we actually follow the advice from the react router team if at all possible:
>Instead of upgrading and updating all of your code at once (which is incredibly difficult and prone to bugs), this strategy enables you to upgrade one component, one hook, and one route at a time by running both v5 and v6 in parallel. Any code you haven't touched is still running the very same code it was before. Once all components are exclusively using the v6 APIs, your app no longer needs the compatibility package and is running on v6.


It was unclear to me from your [comment above](https://github.com/determined-ai/determined/pull/4619#issuecomment-1194191120) whether that was impossible, or just somehow difficult.

Side question: what is the motivation for the upgrade in the first place? is v5 nearing end of support?

keita-determined commented:
I'll fix the issue around password. Thanks for finding

> It was unclear to me from your https://github.com/determined-ai/determined/pull/4619#issuecomment-1194191120 whether that was impossible, or just somehow difficult.

I followed the instruction step by step, but at some point, project got broken. We heavily use `history`, so I assume it might be the reason why it is difficult to upgrade step by step. I won't say it's impossible but IMO it is hard to gradually upgrade.

> Side question: what is the motivation for the upgrade in the first place? is v5 nearing end of support?

The motivation is that lots of libraries around React is not updated for several months. The longer we wait to update libraries, the more difficult it will be in the future. Especially, React 18 and React-Router-Dom v6 have some breaking changes.

keita-determined commented:
I close this PR to start over.

keita-determined commented:
I close this PR to start over.
